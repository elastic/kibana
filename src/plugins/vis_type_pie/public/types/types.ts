/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { Position } from '@elastic/charts';
import { UiCounterMetricType } from '@kbn/analytics';
import { DatatableColumn, SerializedFieldFormat } from '../../../expressions/public';
import { ExpressionValueVisDimension } from '../../../visualizations/public';
import { ExpressionValuePieLabels } from '../expression_functions/pie_labels';
import { PaletteOutput, ChartsPluginSetup } from '../../../charts/public';

export interface Dimension {
  accessor: number;
  format: {
    id?: string;
    params?: SerializedFieldFormat<object>;
  };
}

export interface Dimensions {
  metric: Dimension;
  buckets?: Dimension[];
  splitRow?: Dimension[];
  splitColumn?: Dimension[];
}

interface PieCommonParams {
  addTooltip: boolean;
  addLegend: boolean;
  legendPosition: Position;
  nestedLegend: boolean;
  distinctColors: boolean;
  isDonut: boolean;
}

export interface LabelsParams {
  show: boolean;
  last_level: boolean;
  position: LabelPositions;
  values: boolean;
  truncate: number | null;
  valuesFormat: ValueFormats;
  percentDecimals: number;
}

export interface PieVisParams extends PieCommonParams {
  dimensions: Dimensions;
  labels: LabelsParams;
  palette: PaletteOutput;
}

export interface PieVisConfig extends PieCommonParams {
  buckets?: ExpressionValueVisDimension[];
  metric: ExpressionValueVisDimension;
  splitColumn?: ExpressionValueVisDimension[];
  splitRow?: ExpressionValueVisDimension[];
  labels: ExpressionValuePieLabels;
  palette: string;
}

export interface BucketColumns extends DatatableColumn {
  format?: {
    id?: string;
    params?: SerializedFieldFormat<object>;
  };
}

export enum LabelPositions {
  INSIDE = 'inside',
  DEFAULT = 'default',
}

export enum ValueFormats {
  PERCENT = 'percent',
  VALUE = 'value',
}

export interface PieTypeProps {
  showElasticChartsOptions?: boolean;
  palettes?: ChartsPluginSetup['palettes'];
  trackUiMetric?: (metricType: UiCounterMetricType, eventName: string | string[]) => void;
}

export interface SplitDimensionParams {
  order?: string;
  orderBy?: string;
}

export interface PieContainerDimensions {
  width: number;
  height: number;
}
