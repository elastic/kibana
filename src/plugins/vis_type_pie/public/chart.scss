.pieChart__wrapper,
.pieChart__container {
  display: flex;
  flex: 1 1 auto;
  min-height: 0;
  min-width: 0;
}

.pieChart__container {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  padding: $euiSizeS;
  margin-left: auto;
  margin-right: auto;
}
