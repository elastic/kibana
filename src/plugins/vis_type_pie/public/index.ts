/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { VisTypePiePlugin } from './plugin';

export { pieVisType } from './vis_type';
export { Dimensions, Dimension } from './types';

export const plugin = () => new VisTypePiePlugin();
