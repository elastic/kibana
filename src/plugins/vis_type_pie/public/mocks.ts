/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { Datatable } from '../../expressions/public';
import { BucketColumns, PieVisParams, LabelPositions, ValueFormats } from './types';

export const createMockBucketColumns = (): BucketColumns[] => {
  return [
    {
      id: 'col-0-2',
      name: 'Carrier: Descending',
      meta: {
        type: 'string',
        field: 'Carrier',
        index: 'kibana_sample_data_flights',
        params: {
          id: 'terms',
          params: {
            id: 'string',
            otherBucketLabel: 'Other',
            missingBucketLabel: 'Missing',
          },
        },
        source: 'esaggs',
        sourceParams: {
          indexPatternId: 'd3d7af60-4c81-11e8-b3d7-01146121b73d',
          id: '2',
          enabled: true,
          type: 'terms',
          params: {
            field: 'Carrier',
            orderBy: '1',
            order: 'desc',
            size: 5,
            otherBucket: false,
            otherBucketLabel: 'Other',
            missingBucket: false,
            missingBucketLabel: 'Missing',
          },
          schema: 'segment',
        },
      },
      format: {
        id: 'terms',
        params: {
          id: 'string',
        },
      },
    },
    {
      id: 'col-2-3',
      name: 'Cancelled: Descending',
      meta: {
        type: 'boolean',
        field: 'Cancelled',
        index: 'kibana_sample_data_flights',
        params: {
          id: 'terms',
          params: {
            id: 'boolean',
            otherBucketLabel: 'Other',
            missingBucketLabel: 'Missing',
          },
        },
        source: 'esaggs',
        sourceParams: {
          indexPatternId: 'd3d7af60-4c81-11e8-b3d7-01146121b73d',
          id: '3',
          enabled: true,
          type: 'terms',
          params: {
            field: 'Cancelled',
            orderBy: '1',
            order: 'desc',
            size: 5,
            otherBucket: false,
            otherBucketLabel: 'Other',
            missingBucket: false,
            missingBucketLabel: 'Missing',
          },
          schema: 'segment',
        },
      },
      format: {
        id: 'terms',
        params: {
          id: 'boolean',
        },
      },
    },
  ];
};

export const createMockVisData = (): Datatable => {
  return {
    type: 'datatable',
    rows: [
      {
        'col-0-2': 'Logstash Airways',
        'col-2-3': 0,
        'col-1-1': 797,
        'col-3-1': 689,
      },
      {
        'col-0-2': 'Logstash Airways',
        'col-2-3': 1,
        'col-1-1': 797,
        'col-3-1': 108,
      },
      {
        'col-0-2': 'JetBeats',
        'col-2-3': 0,
        'col-1-1': 766,
        'col-3-1': 654,
      },
      {
        'col-0-2': 'JetBeats',
        'col-2-3': 1,
        'col-1-1': 766,
        'col-3-1': 112,
      },
      {
        'col-0-2': 'ES-Air',
        'col-2-3': 0,
        'col-1-1': 744,
        'col-3-1': 665,
      },
      {
        'col-0-2': 'ES-Air',
        'col-2-3': 1,
        'col-1-1': 744,
        'col-3-1': 79,
      },
      {
        'col-0-2': 'Kibana Airlines',
        'col-2-3': 0,
        'col-1-1': 731,
        'col-3-1': 655,
      },
      {
        'col-0-2': 'Kibana Airlines',
        'col-2-3': 1,
        'col-1-1': 731,
        'col-3-1': 76,
      },
    ],
    columns: [
      {
        id: 'col-0-2',
        name: 'Carrier: Descending',
        meta: {
          type: 'string',
          field: 'Carrier',
          index: 'kibana_sample_data_flights',
          params: {
            id: 'terms',
            params: {
              id: 'string',
              otherBucketLabel: 'Other',
              missingBucketLabel: 'Missing',
            },
          },
          source: 'esaggs',
          sourceParams: {
            indexPatternId: 'd3d7af60-4c81-11e8-b3d7-01146121b73d',
            id: '2',
            enabled: true,
            type: 'terms',
            params: {
              field: 'Carrier',
              orderBy: '1',
              order: 'desc',
              size: 5,
              otherBucket: false,
              otherBucketLabel: 'Other',
              missingBucket: false,
              missingBucketLabel: 'Missing',
            },
            schema: 'segment',
          },
        },
      },
      {
        id: 'col-1-1',
        name: 'Count',
        meta: {
          type: 'number',
          index: 'kibana_sample_data_flights',
          params: {
            id: 'number',
          },
          source: 'esaggs',
          sourceParams: {
            indexPatternId: 'd3d7af60-4c81-11e8-b3d7-01146121b73d',
            id: '1',
            enabled: true,
            type: 'count',
            params: {},
            schema: 'metric',
          },
        },
      },
      {
        id: 'col-2-3',
        name: 'Cancelled: Descending',
        meta: {
          type: 'boolean',
          field: 'Cancelled',
          index: 'kibana_sample_data_flights',
          params: {
            id: 'terms',
            params: {
              id: 'boolean',
              otherBucketLabel: 'Other',
              missingBucketLabel: 'Missing',
            },
          },
          source: 'esaggs',
          sourceParams: {
            indexPatternId: 'd3d7af60-4c81-11e8-b3d7-01146121b73d',
            id: '3',
            enabled: true,
            type: 'terms',
            params: {
              field: 'Cancelled',
              orderBy: '1',
              order: 'desc',
              size: 5,
              otherBucket: false,
              otherBucketLabel: 'Other',
              missingBucket: false,
              missingBucketLabel: 'Missing',
            },
            schema: 'segment',
          },
        },
      },
      {
        id: 'col-3-1',
        name: 'Count',
        meta: {
          type: 'number',
          index: 'kibana_sample_data_flights',
          params: {
            id: 'number',
          },
          source: 'esaggs',
          sourceParams: {
            indexPatternId: 'd3d7af60-4c81-11e8-b3d7-01146121b73d',
            id: '1',
            enabled: true,
            type: 'count',
            params: {},
            schema: 'metric',
          },
        },
      },
    ],
  };
};

export const createMockPieParams = (): PieVisParams => {
  return ({
    addLegend: true,
    addTooltip: true,
    isDonut: true,
    labels: {
      position: LabelPositions.DEFAULT,
      show: true,
      truncate: 100,
      values: true,
      valuesFormat: ValueFormats.PERCENT,
      percentDecimals: 2,
    },
    legendPosition: 'right',
    nestedLegend: false,
    distinctColors: false,
    palette: {
      name: 'default',
      type: 'palette',
    },
    type: 'pie',
    dimensions: {
      metric: {
        accessor: 1,
        format: {
          id: 'number',
        },
        params: {},
        label: 'Count',
        aggType: 'count',
      },
      buckets: [
        {
          accessor: 0,
          format: {
            id: 'terms',
            params: {
              id: 'string',
              otherBucketLabel: 'Other',
              missingBucketLabel: 'Missing',
            },
          },
          label: 'Carrier: Descending',
          aggType: 'terms',
        },
        {
          accessor: 2,
          format: {
            id: 'terms',
            params: {
              id: 'boolean',
              otherBucketLabel: 'Other',
              missingBucketLabel: 'Missing',
            },
          },
          label: 'Cancelled: Descending',
          aggType: 'terms',
        },
      ],
    },
  } as unknown) as PieVisParams;
};
