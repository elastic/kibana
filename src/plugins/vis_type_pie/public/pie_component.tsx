/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, { memo, useCallback, useMemo, useState, useEffect, useRef } from 'react';

import {
  Chart,
  Datum,
  LayerValue,
  Partition,
  Position,
  Settings,
  RenderChangeListener,
  TooltipProps,
  TooltipType,
  SeriesIdentifier,
} from '@elastic/charts';
import {
  LegendToggle,
  ClickTriggerEvent,
  ChartsPluginSetup,
  PaletteRegistry,
} from '../../charts/public';
import { DataPublicPluginStart, FieldFormat } from '../../data/public';
import type { PersistedState } from '../../visualizations/public';
import { Datatable, DatatableColumn, IInterpreterRenderHandlers } from '../../expressions/public';
import { DEFAULT_PERCENT_DECIMALS } from '../common';
import { PieVisParams, BucketColumns, ValueFormats, PieContainerDimensions } from './types';
import {
  getColorPicker,
  getLayers,
  getLegendActions,
  canFilter,
  getFilterClickData,
  getFilterEventData,
  getConfig,
  getColumns,
  getSplitDimensionAccessor,
} from './utils';
import { ChartSplit, SMALL_MULTIPLES_ID } from './components/chart_split';

import './chart.scss';

declare global {
  interface Window {
    /**
     * Flag used to enable debugState on elastic charts
     */
    _echDebugStateFlag?: boolean;
  }
}

export interface PieComponentProps {
  visParams: PieVisParams;
  visData: Datatable;
  uiState: PersistedState;
  fireEvent: IInterpreterRenderHandlers['event'];
  renderComplete: IInterpreterRenderHandlers['done'];
  chartsThemeService: ChartsPluginSetup['theme'];
  palettesRegistry: PaletteRegistry;
  services: DataPublicPluginStart;
  syncColors: boolean;
}

const PieComponent = (props: PieComponentProps) => {
  const chartTheme = props.chartsThemeService.useChartsTheme();
  const chartBaseTheme = props.chartsThemeService.useChartsBaseTheme();
  const [showLegend, setShowLegend] = useState<boolean>(() => {
    const bwcLegendStateDefault =
      props.visParams.addLegend == null ? false : props.visParams.addLegend;
    return props.uiState?.get('vis.legendOpen', bwcLegendStateDefault) as boolean;
  });
  const [dimensions, setDimensions] = useState<undefined | PieContainerDimensions>();

  const parentRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (parentRef && parentRef.current) {
      const parentHeight = parentRef.current!.getBoundingClientRect().height;
      const parentWidth = parentRef.current!.getBoundingClientRect().width;
      setDimensions({ width: parentWidth, height: parentHeight });
    }
  }, [parentRef]);

  const onRenderChange = useCallback<RenderChangeListener>(
    (isRendered) => {
      if (isRendered) {
        props.renderComplete();
      }
    },
    [props]
  );

  // handles slice click event
  const handleSliceClick = useCallback(
    (
      clickedLayers: LayerValue[],
      bucketColumns: Array<Partial<BucketColumns>>,
      visData: Datatable,
      splitChartDimension?: DatatableColumn,
      splitChartFormatter?: FieldFormat
    ): void => {
      const data = getFilterClickData(
        clickedLayers,
        bucketColumns,
        visData,
        splitChartDimension,
        splitChartFormatter
      );
      const event = {
        name: 'filterBucket',
        data: { data },
      };
      props.fireEvent(event);
    },
    [props]
  );

  // handles legend action event data
  const getLegendActionEventData = useCallback(
    (visData: Datatable) => (series: SeriesIdentifier): ClickTriggerEvent | null => {
      const data = getFilterEventData(visData, series);

      return {
        name: 'filterBucket',
        data: {
          negate: false,
          data,
        },
      };
    },
    []
  );

  const handleLegendAction = useCallback(
    (event: ClickTriggerEvent, negate = false) => {
      props.fireEvent({
        ...event,
        data: {
          ...event.data,
          negate,
        },
      });
    },
    [props]
  );

  const toggleLegend = useCallback(() => {
    setShowLegend((value) => {
      const newValue = !value;
      props.uiState?.set('vis.legendOpen', newValue);
      return newValue;
    });
  }, [props.uiState]);

  useEffect(() => {
    setShowLegend(props.visParams.addLegend);
    props.uiState?.set('vis.legendOpen', props.visParams.addLegend);
  }, [props.uiState, props.visParams.addLegend]);

  const setColor = useCallback(
    (newColor: string | null, seriesLabel: string | number) => {
      const colors = props.uiState?.get('vis.colors') || {};
      if (colors[seriesLabel] === newColor || !newColor) {
        delete colors[seriesLabel];
      } else {
        colors[seriesLabel] = newColor;
      }
      props.uiState?.setSilent('vis.colors', null);
      props.uiState?.set('vis.colors', colors);
      props.uiState?.emit('reload');
    },
    [props.uiState]
  );

  const { visData, visParams, services, syncColors } = props;

  function getSliceValue(d: Datum, metricColumn: DatatableColumn) {
    if (typeof d[metricColumn.id] === 'number' && d[metricColumn.id] !== 0) {
      return d[metricColumn.id];
    }
    return Number.EPSILON;
  }

  // formatters
  const metricFieldFormatter = services.fieldFormats.deserialize(
    visParams.dimensions.metric.format
  );
  const splitChartFormatter = visParams.dimensions.splitColumn
    ? services.fieldFormats.deserialize(visParams.dimensions.splitColumn[0].format)
    : visParams.dimensions.splitRow
    ? services.fieldFormats.deserialize(visParams.dimensions.splitRow[0].format)
    : undefined;
  const percentFormatter = services.fieldFormats.deserialize({
    id: 'percent',
    params: {
      pattern: `0,0.[${'0'.repeat(visParams.labels.percentDecimals ?? DEFAULT_PERCENT_DECIMALS)}]%`,
    },
  });

  const { bucketColumns, metricColumn } = useMemo(() => getColumns(visParams, visData), [
    visData,
    visParams,
  ]);

  const layers = useMemo(
    () =>
      getLayers(
        bucketColumns,
        visParams,
        props.uiState?.get('vis.colors', {}),
        visData.rows,
        props.palettesRegistry,
        services.fieldFormats,
        syncColors
      ),
    [
      bucketColumns,
      visParams,
      props.uiState,
      props.palettesRegistry,
      visData.rows,
      services.fieldFormats,
      syncColors,
    ]
  );
  const config = useMemo(() => getConfig(visParams, chartTheme, dimensions), [
    chartTheme,
    visParams,
    dimensions,
  ]);
  const tooltip: TooltipProps = {
    type: visParams.addTooltip ? TooltipType.Follow : TooltipType.None,
  };
  const legendPosition = visParams.legendPosition ?? Position.Right;

  const legendColorPicker = useMemo(
    () =>
      getColorPicker(
        legendPosition,
        setColor,
        bucketColumns,
        visParams.palette.name,
        visData.rows,
        props.uiState,
        visParams.distinctColors
      ),
    [
      legendPosition,
      setColor,
      bucketColumns,
      visParams.palette.name,
      visParams.distinctColors,
      visData.rows,
      props.uiState,
    ]
  );

  const splitChartColumnAccessor = visParams.dimensions.splitColumn
    ? getSplitDimensionAccessor(
        services.fieldFormats,
        visData.columns
      )(visParams.dimensions.splitColumn[0])
    : undefined;
  const splitChartRowAccessor = visParams.dimensions.splitRow
    ? getSplitDimensionAccessor(
        services.fieldFormats,
        visData.columns
      )(visParams.dimensions.splitRow[0])
    : undefined;

  const splitChartDimension = visParams.dimensions.splitColumn
    ? visData.columns[visParams.dimensions.splitColumn[0].accessor]
    : visParams.dimensions.splitRow
    ? visData.columns[visParams.dimensions.splitRow[0].accessor]
    : undefined;

  return (
    <div className="pieChart__container" data-test-subj="visTypePieChart">
      <div className="pieChart__wrapper" ref={parentRef}>
        <LegendToggle
          onClick={toggleLegend}
          showLegend={showLegend}
          legendPosition={legendPosition}
        />
        <Chart size="100%">
          <ChartSplit
            splitColumnAccessor={splitChartColumnAccessor}
            splitRowAccessor={splitChartRowAccessor}
            splitDimension={splitChartDimension}
          />
          <Settings
            debugState={window._echDebugStateFlag ?? false}
            showLegend={showLegend}
            legendPosition={legendPosition}
            legendMaxDepth={visParams.nestedLegend ? undefined : 1}
            legendColorPicker={legendColorPicker}
            flatLegend={Boolean(splitChartDimension)}
            tooltip={tooltip}
            onElementClick={(args) => {
              handleSliceClick(
                args[0][0] as LayerValue[],
                bucketColumns,
                visData,
                splitChartDimension,
                splitChartFormatter
              );
            }}
            legendAction={getLegendActions(
              canFilter,
              getLegendActionEventData(visData),
              handleLegendAction,
              visParams,
              services.actions,
              services.fieldFormats
            )}
            theme={chartTheme}
            baseTheme={chartBaseTheme}
            onRenderChange={onRenderChange}
          />
          <Partition
            id="pie"
            smallMultiples={SMALL_MULTIPLES_ID}
            data={visData.rows}
            valueAccessor={(d: Datum) => getSliceValue(d, metricColumn)}
            percentFormatter={(d: number) => percentFormatter.convert(d / 100)}
            valueGetter={
              !visParams.labels.show ||
              visParams.labels.valuesFormat === ValueFormats.VALUE ||
              !visParams.labels.values
                ? undefined
                : 'percent'
            }
            valueFormatter={(d: number) =>
              !visParams.labels.show || !visParams.labels.values
                ? ''
                : metricFieldFormatter.convert(d)
            }
            layers={layers}
            config={config}
            topGroove={!visParams.labels.show ? 0 : undefined}
          />
        </Chart>
      </div>
    </div>
  );
};

// eslint-disable-next-line import/no-default-export
export default memo(PieComponent);
