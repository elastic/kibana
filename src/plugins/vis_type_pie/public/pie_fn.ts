/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { i18n } from '@kbn/i18n';
import { ExpressionFunctionDefinition, Datatable, Render } from '../../expressions/public';
import { PieVisParams, PieVisConfig } from './types';

export const vislibPieName = 'pie_vis';

export interface RenderValue {
  visData: Datatable;
  visType: string;
  visConfig: PieVisParams;
  syncColors: boolean;
}

export type VisTypePieExpressionFunctionDefinition = ExpressionFunctionDefinition<
  typeof vislibPieName,
  Datatable,
  PieVisConfig,
  Render<RenderValue>
>;

export const createPieVisFn = (): VisTypePieExpressionFunctionDefinition => ({
  name: vislibPieName,
  type: 'render',
  inputTypes: ['datatable'],
  help: i18n.translate('visTypePie.functions.help', {
    defaultMessage: 'Pie visualization',
  }),
  args: {
    metric: {
      types: ['vis_dimension'],
      help: i18n.translate('visTypePie.function.args.metricHelpText', {
        defaultMessage: 'Metric dimensions config',
      }),
      required: true,
    },
    buckets: {
      types: ['vis_dimension'],
      help: i18n.translate('visTypePie.function.args.bucketsHelpText', {
        defaultMessage: 'Buckets dimensions config',
      }),
      multi: true,
    },
    splitColumn: {
      types: ['vis_dimension'],
      help: i18n.translate('visTypePie.function.args.splitColumnHelpText', {
        defaultMessage: 'Split by column dimension config',
      }),
      multi: true,
    },
    splitRow: {
      types: ['vis_dimension'],
      help: i18n.translate('visTypePie.function.args.splitRowHelpText', {
        defaultMessage: 'Split by row dimension config',
      }),
      multi: true,
    },
    addTooltip: {
      types: ['boolean'],
      help: i18n.translate('visTypePie.function.args.addTooltipHelpText', {
        defaultMessage: 'Show tooltip on slice hover',
      }),
      default: true,
    },
    addLegend: {
      types: ['boolean'],
      help: i18n.translate('visTypePie.function.args.addLegendHelpText', {
        defaultMessage: 'Show legend chart legend',
      }),
    },
    legendPosition: {
      types: ['string'],
      help: i18n.translate('visTypePie.function.args.legendPositionHelpText', {
        defaultMessage: 'Position the legend on top, bottom, left, right of the chart',
      }),
    },
    nestedLegend: {
      types: ['boolean'],
      help: i18n.translate('visTypePie.function.args.nestedLegendHelpText', {
        defaultMessage: 'Show a more detailed legend',
      }),
      default: false,
    },
    distinctColors: {
      types: ['boolean'],
      help: i18n.translate('visTypePie.function.args.distinctColorsHelpText', {
        defaultMessage:
          'Maps different color per slice. Slices with the same value have the same color',
      }),
      default: false,
    },
    isDonut: {
      types: ['boolean'],
      help: i18n.translate('visTypePie.function.args.isDonutHelpText', {
        defaultMessage: 'Displays the pie chart as donut',
      }),
      default: false,
    },
    palette: {
      types: ['string'],
      help: i18n.translate('visTypePie.function.args.paletteHelpText', {
        defaultMessage: 'Defines the chart palette name',
      }),
      default: 'default',
    },
    labels: {
      types: ['pie_labels'],
      help: i18n.translate('visTypePie.function.args.labelsHelpText', {
        defaultMessage: 'Pie labels config',
      }),
    },
  },
  fn(context, args, handlers) {
    const visConfig = {
      ...args,
      palette: {
        type: 'palette',
        name: args.palette,
      },
      dimensions: {
        metric: args.metric,
        buckets: args.buckets,
        splitColumn: args.splitColumn,
        splitRow: args.splitRow,
      },
    } as PieVisParams;

    if (handlers?.inspectorAdapters?.tables) {
      handlers.inspectorAdapters.tables.logDatatable('default', context);
    }

    return {
      type: 'render',
      as: vislibPieName,
      value: {
        visData: context,
        visConfig,
        syncColors: handlers?.isSyncColorsEnabled?.() ?? false,
        visType: 'pie',
        params: {
          listenOnChange: true,
        },
      },
    };
  },
});
