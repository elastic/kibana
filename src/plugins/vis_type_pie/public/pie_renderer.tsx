/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, { lazy } from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { I18nProvider } from '@kbn/i18n/react';
import { ExpressionRenderDefinition } from '../../expressions/public';
import { VisualizationContainer } from '../../visualizations/public';
import type { PersistedState } from '../../visualizations/public';
import { VisTypePieDependencies } from './plugin';

import { RenderValue, vislibPieName } from './pie_fn';

const PieComponent = lazy(() => import('./pie_component'));

function shouldShowNoResultsMessage(visData: any): boolean {
  const rows: object[] | undefined = visData?.rows;
  const isZeroHits = visData?.hits === 0 || (rows && !rows.length);

  return Boolean(isZeroHits);
}

export const getPieVisRenderer: (
  deps: VisTypePieDependencies
) => ExpressionRenderDefinition<RenderValue> = ({ theme, palettes, getStartDeps }) => ({
  name: vislibPieName,
  displayName: 'Pie visualization',
  reuseDomNode: true,
  render: async (domNode, { visConfig, visData, syncColors }, handlers) => {
    const showNoResult = shouldShowNoResultsMessage(visData);

    handlers.onDestroy(() => {
      unmountComponentAtNode(domNode);
    });

    const services = await getStartDeps();
    const palettesRegistry = await palettes.getPalettes();

    render(
      <I18nProvider>
        <VisualizationContainer handlers={handlers} showNoResult={showNoResult}>
          <PieComponent
            chartsThemeService={theme}
            palettesRegistry={palettesRegistry}
            visParams={visConfig}
            visData={visData}
            renderComplete={handlers.done}
            fireEvent={handlers.event}
            uiState={handlers.uiState as PersistedState}
            services={services.data}
            syncColors={syncColors}
          />
        </VisualizationContainer>
      </I18nProvider>,
      domNode
    );
  },
});
