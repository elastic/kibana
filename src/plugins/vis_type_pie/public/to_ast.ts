/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { getVisSchemas, VisToExpressionAst, SchemaConfig } from '../../visualizations/public';
import { buildExpression, buildExpressionFunction } from '../../expressions/public';
import { PieVisParams, LabelsParams } from './types';
import { vislibPieName, VisTypePieExpressionFunctionDefinition } from './pie_fn';
import { getEsaggsFn } from './to_ast_esaggs';

const prepareDimension = (params: SchemaConfig) => {
  const visdimension = buildExpressionFunction('visdimension', { accessor: params.accessor });

  if (params.format) {
    visdimension.addArgument('format', params.format.id);
    visdimension.addArgument('formatParams', JSON.stringify(params.format.params));
  }

  return buildExpression([visdimension]);
};

const prepareLabels = (params: LabelsParams) => {
  const pieLabels = buildExpressionFunction('pielabels', {
    show: params.show,
    lastLevel: params.last_level,
    values: params.values,
    truncate: params.truncate,
  });
  if (params.position) {
    pieLabels.addArgument('position', params.position);
  }
  if (params.valuesFormat) {
    pieLabels.addArgument('valuesFormat', params.valuesFormat);
  }
  if (params.percentDecimals != null) {
    pieLabels.addArgument('percentDecimals', params.percentDecimals);
  }
  return buildExpression([pieLabels]);
};

export const toExpressionAst: VisToExpressionAst<PieVisParams> = async (vis, params) => {
  const schemas = getVisSchemas(vis, params);
  const args = {
    // explicitly pass each param to prevent extra values trapping
    addTooltip: vis.params.addTooltip,
    addLegend: vis.params.addLegend,
    legendPosition: vis.params.legendPosition,
    nestedLegend: vis.params?.nestedLegend,
    distinctColors: vis.params?.distinctColors,
    isDonut: vis.params.isDonut,
    palette: vis.params?.palette?.name,
    labels: prepareLabels(vis.params.labels),
    metric: schemas.metric.map(prepareDimension),
    buckets: schemas.segment?.map(prepareDimension),
    splitColumn: schemas.split_column?.map(prepareDimension),
    splitRow: schemas.split_row?.map(prepareDimension),
  };

  const visTypePie = buildExpressionFunction<VisTypePieExpressionFunctionDefinition>(
    vislibPieName,
    args
  );

  const ast = buildExpression([getEsaggsFn(vis), visTypePie]);

  return ast.toAst();
};
