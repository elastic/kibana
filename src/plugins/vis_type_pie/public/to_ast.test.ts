/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { Vis } from '../../visualizations/public';

import { PieVisParams } from './types';
import { samplePieVis } from './sample_vis.test.mocks';
import { toExpressionAst } from './to_ast';

describe('vis type pie vis toExpressionAst function', () => {
  let vis: Vis<PieVisParams>;
  const params = {
    timefilter: {},
    timeRange: {},
    abortSignal: {},
  } as any;

  beforeEach(() => {
    vis = samplePieVis as any;
  });

  it('should match basic snapshot', async () => {
    const actual = await toExpressionAst(vis, params);
    expect(actual).toMatchSnapshot();
  });
});
