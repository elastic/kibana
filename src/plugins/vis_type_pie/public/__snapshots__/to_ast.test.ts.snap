// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pie vis toExpressionAst function should match basic snapshot 1`] = `undefined`;
